### DISCUSSION TITLE
Forward reference for recursive models.

### DISCUSSION THREAD

**Discussion by goraj** [fastapi/fastapi#7898] [Questions] [answered]
### First check

* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.

### Description

Hello,
I have a recursive model of the following form using a forward reference:
```
class FilterGroup(BaseModel):
    filters: List[Union[Filter, 'FilterGroup']]
    operation: Literal['and_', 'or_']
```

Understandably the schema generation fails for it. Is there a workaround to use this as input in fastapi?

Thank you very much.

**Comment by goraj** [comment]
You can use the following workaround (Thanks @ludwig404): 
```
from __future__ import annotations
from typing import List, Union
from pydantic import BaseModel

class Filter(BaseModel):
    kind: str

# Ducktyping to support recursive model for FilterGroup.
class FilterGroupDuck(BaseModel):
    filters: List[Union[Filter, FilterGroup]]
    operation: Literal['and_', 'or_']

class FilterGroup(BaseModel):
    filters: List[Union[Filter, FilterGroupDuck]]
    operation: Literal['and_', 'or_']

FilterGroupDuck.update_forward_refs()
FilterGroup.update_forward_refs()
```

**Comment by tiangolo** [comment]
Thanks for reporting back and closing the issue :+1:

**Comment by connebs** [comment, marked as answer]
I don't actually think you need to do the whole duck-typing thing, using a string reference and then updating the refs should be all you need, e.g.

```py
from typing import List, Union
from pydantic import BaseModel

class Filter(BaseModel):
    kind: str

class FilterGroup(BaseModel):
    filters: List[Union[Filter, "FilterGroup"]]
    operation: Literal['and_', 'or_']

FilterGroup.update_forward_refs()
```

**Comment by kfuerst** [comment]
Adopting the answer from acnebs to reference FilterGroup as class:

from typing import List, Union, ForwardRef
from pydantic import BaseModel

class Filter(BaseModel):
    kind: str

FilterGroup = ForwardRef('FilterGroup')

class FilterGroup(BaseModel):
    filters: List[Union[Filter, FilterGroup]]
    operation: Literal['and_', 'or_']

FilterGroup.update_forward_refs()
